connect/service.go

```diff
@@ -53,29 +53,20 @@ type Service struct {
 	logger hclog.Logger
 }
 
-// NewService creates and starts a Service. The caller must close the returned
-// service to free resources and allow the program to exit normally. This is
-// typically called in a signal handler.
-//
-// Caller must provide client which is already configured to speak to the local
-// Consul agent, and with an ACL token that has `service:write` privileges for
-// the service specified.
-func NewService(serviceName string, client *api.Client) (*Service, error) {
-	logger := hclog.New(&hclog.LoggerOptions{})
-
-	return NewServiceWithLogger(serviceName, client,
-		logger)
+type Config struct {
+	Client           *api.Client
+	Logger           hclog.Logger
+	ServerNextProtos *[]string
 }
 
-// NewServiceWithLogger starts the service with a specified log.Logger.
-func NewServiceWithLogger(serviceName string, client *api.Client,
-	logger hclog.Logger) (*Service, error) {
+// NewServiceWithConfig starts a service with the specified Config.
+func NewServiceWithConfig(serviceName string, config Config) (*Service, error) {
 	s := &Service{
 		service:              serviceName,
-		client:               client,
-		logger:               logger.Named(logging.Connect).With("service", serviceName),
-		tlsCfg:               newDynamicTLSConfig(defaultTLSConfig(), logger),
-		httpResolverFromAddr: ConsulResolverFromAddrFunc(client),
+		client:               config.Client,
+		logger:               config.Logger.Named(logging.Connect).With("service", serviceName),
```

Comment:
I think it would be friendlier to gracefully handle a nil logger in the Config in the same way we do in the other `New*` funcs. If we moved that defaulting behaviour here it would only be needed in one place and not both here and `NewService`.

Reply:
So this would mean I'd switch the config struct to have a pointer to a logger? Because otherwise it cannot be `nil` or is my go knowledge this bad?